File: Algorithm_18_4Sum.py
class Solution:
    def fourSum(self, nums, target):
        tmp,res=[],[]
        nums.sort()
        for i in range(len(nums)-3):
            if i>0 and nums[i]==nums[i-1]: continue
            if nums[i]+nums[i+1]+nums[i+2]+nums[i+3]>target: break
            if nums[i]+nums[len(nums)-1]+nums[len(nums)-2]+nums[len(nums)-3]<target: continue
            for j in range(i+1,len(nums)-2):
                if j>i+1 and nums[j]==nums[j-1]:continue
                if nums[j]+nums[j+1]+nums[j+2]>target-nums[i]: break
                if nums[j]+nums[len(nums)-1]+nums[len(nums)-2]<target-nums[i]: continue
                low,high=j+1,len(nums)-1
                while low<high:
                    if low<high and nums[i]+nums[j]+nums[low]+nums[high]==target:
                        tmp.append([nums[i],nums[j],nums[low],nums[high]])
                        while low<high and nums[low]==nums[low+1]: low+=1
                        while low<high and nums[high]==nums[high-1]: high-=1
                        low+=1
                        high-=1
                    elif nums[i]+nums[j]+nums[low]+nums[high]<target: low+=1
                    else: high-=1
        for i in range(len(tmp)):
            if res.count(tmp[i])==0: res.append(tmp[i])
        return res

File: Algorithm_18_4Sum.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

// 思路：与Algorithm_15_3Sum相同，先排序，然后保持i、j不变，移动left和right
// 注意消除重复

class Solution {

    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 3; i++) {
            for (int j = i + 1; j < nums.length - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) { // avoid duplicate for nums[j]
                    continue;
                }
                if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target) { // too large
                    break;
                }
                if (nums[i] + nums[j] + nums[nums.length - 1] + nums[nums.length - 2] < target) { // too small
                    continue;
                }
                int left = j + 1, right = nums.length - 1;
                while (left < right) {
                    if (nums[i] + nums[j] + nums[left] + nums[right] == target) {
                        res.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        while (left < right && nums[left + 1] == nums[left]) { // avoid duplicate for nums[left]
                            left++;
                        }
                        while (left < right && nums[right - 1] == nums[right]) { // avoid duplicate for nums[right]
                            right--;
                        }
                        left++;
                        right--;
                    } else if (nums[i] + nums[j] + nums[left] + nums[right] < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
            while (i < nums.length - 4 && nums[i + 1] == nums[i]) { // avoid duplicate for nums[i]
                i++;
            }
        }
        return res;
    }

}

// k-Sum问题通用解法

public class GeneralSolution {
    int len = 0;
    public List<List<Integer>> fourSum(int[] nums, int target) {
        len = nums.length;
        Arrays.sort(nums);
        return kSum(nums, target, 4, 0);
    }
   private ArrayList<List<Integer>> kSum(int[] nums, int target, int k, int index) {
        ArrayList<List<Integer>> res = new ArrayList<List<Integer>>();
        if(index >= len) {
            return res;
        }
        if(k == 2) {
            int i = index, j = len - 1;
            while(i < j) {
                //find a pair
                if(target - nums[i] == nums[j]) {
                    List<Integer> temp = new ArrayList<>();
                    temp.add(nums[i]);
                    temp.add(target-nums[i]);
                    res.add(temp);
                    //skip duplication
                    while(i<j && nums[i]==nums[i+1]) i++;
                    while(i<j && nums[j-1]==nums[j]) j--;
                    i++;
                    j--;
                //move left bound
                } else if (target - nums[i] > nums[j]) {
                    i++;
                //move right bound
                } else {
                    j--;
                }
            }
        } else {
            for (int i = index; i < len - k + 1; i++) {
                //use current number to reduce ksum into k-1sum
                ArrayList<List<Integer>> temp = kSum(nums, target - nums[i], k-1, i+1);
                if(temp != null){
                    //add previous results
                    for (List<Integer> t : temp) {
                        t.add(0, nums[i]);
                    }
                    res.addAll(temp);
                }
                while (i < len - 1 && nums[i] == nums[i+1]) {
                    //skip duplicated numbers
                    i++;
                }
            }
        }
        return res;
    }
}